### DISCUSSION TITLE
`clipPath` related properties

### DISCUSSION THREAD

**Discussion by ShaMan123** [fabricjs/fabric.js#7138] [Ideas]
I think the props `inverted` and `absolutePositioned` should be changed. 
They are extremely unclear and confusing. Their names and the fact that they are set on the clip path object instead of on the object you want to clip.
I suggest unifying them under `clipPathOptions` or `clipPathConfig` and for this prop to reside on the object you want to clip.
This is IMO much more clear and intuitive.
Consider this:
```ts
obj.set({
  clipPath: otherObj,
  clipPathOptions: { inverted: true }
})

``` 

@asturur @melchiar

**Comment by melchiar** [comment]
I think you're right, that probably would have made it more clear. The problem is that making such a change now would break a lot of current applications while not really introducing any new functionality, aside from perhaps adding the ability for each object using a particular clipPath to have different properties (I'm not sure how useful that would be). 

Perhaps there's a functional reason for the clipping properties being attached to the clipPath itself that I'm not aware of. @asturur would be the better person to answer that.

**Reply by ShaMan123** [reply]
The docs state that `absolutePositioned` may change.
http://fabricjs.com/docs/fabric.Object.html#absolutePositioned

In my opinion breaking changes are a must, or else you get a bloated maze of backward compatible nonsense. 
A major version is used exactly for that.

**Reply by asturur** [reply]
Yes absolutePositioned is a feature i deeply regret, is not that much useful. It was supposed to fix some use cases, that at the end are specifc of application like a multi photo frame in which you are able to multiple pictures on multiple spaces but each picture stays in its space. But other than that is pretty much non useful. Is not hard to mantain, but as you have seen with groups is unclear how it should behave.

**Comment by asturur** [comment]
Regarding nested properties in objects, my concerns are the following, and they are mostly about mutability and extra code to handle them.

I see those cons with nested properties:
- using set('property', value) needs extra handling
- using the stateful cache or the stateful object:modified ( i do not like that feature at all anyway) requires extra handling for nested properties
- cloning an object requires more code because you can't just assign props ( this is sort of handled by our custom clone thing )
- restoring fromObject requires extra attention or you end up modifying the original object you are restoring  ( mutability )
- outputing toObject requires extra attention or a change to the exported object can eventually modify the instance ( mutability )
- other generic mutability issues devs can end up with ( like creating the options in an object, assign to 2 fabric instances, and then not knowing those are linked )
- in general is a pattern we don't have elsewhere, is new.

While the pro for me are just:
- it seems more organized and well named

if we want them to be more clear of what they do, we can change the names with a prefix, like,
`clipping_inverted`, `clipping_absolutepositioned`
